Fix versionning²

pierre-24 opened the pull request:
| Q | R |
| --- | --- |
| Correction de bugs ? | oui (!) |
| Nouvelle Fonctionnalité ? | non |
| Tickets concernés | #1189, #1187, #1281 |

(moi aussi, je sais faire des PR avec des titres de mathématiciens :p )

**Description**

Cette PR règle plusieurs problèmes de _versionning_ plus ou moins poussés : 
- 500 sur la liste des tutoriels en béta d'un utilisateur
- Titre des tutoriels non versionné sur une version béta
- Suppression d'un chapitre en _draft_ qui impacte la version _online_

Pour ce faire, je me suis assez fortement inspiré de la fonction `load_dic()` de Firm1 sur sa PR (#1314), rendons à césar ce qui est à lui, avec cependant le chargement de l'intro et de la conclusion que j'ai extradé hors de cette fonction pour des raisons de performances (la plupart du temps, quand `load_dic()` est employé, c'est pour récupérer le titre de la version versionnée, et non son introduction, donc ça sert à rien de la charger). C'est cette dernière modification qui pourrait introduire des régressions.

Logiquement, y'a des tests unitaires pour verouiller le machin, mais qui peut prévoir toutes les possibilités?
# Note pour QA

Forcément; _checker_ tout le workflow des tutos. Plus précisément : 
- Créer un big-tuto nommé "big tuto", avec une partie "partie 1" qui contient un chapitre "chapitre 1", avec un extrait "extrait 1", et "chapitre 2", et une partie "partie 2" qui contient un chapitre "chapitre 3"
- Passer le tutoriel en béta, le parcourir pour vérifier que tout est bien accessible (partie et chapitre)
- Passer le tutoriel en ligne, vérifier que tout est bien accessible (partie et chapitre)
- Modifier tout les titres en leur rajoutant "(modifié)" (donc "big tuto" → "big tuto (modifié)", et ainsi de suite), et changer le texte de "extrait 1". Checker sur la béta que rien n'as changé (dans partie et chapitre AUSSI, bien regarder l'extrait). Checker en ligne que c'est également le cas (donc sur le tuto lui-même, mais aussi sur la page d'accueil, la liste des tutoriels, la liste des tutoriels de l'utilisateur et sa liste béta).
- Mettre à jour la béta, vérifier que cette fois, les titres ont changé sur celle-ci.
- Y aller franchement : supprimer la "partie 2 (modifié)" et le "chapitre 1 (modifié)". Dans la "partie 1 (modifié)", rajouter un "chapitre 4".
- Vérifier sur la béta que le "chapitre 4" n'est pas apparu et que la partie et le chapitre supprimé en _draft_ sont bien présents et accessibles (avec le bon titre et les bons textes, si possible). Faire la même vérification en ligne. 

Adapter ces tests pour les mini-tutoriels (donc juste regarder au titre et au extraits).

Ne pas hésiter à en faire plus et a reporter toute anomalie.

Les utilisateurs du ZdS vous remercient d'avance :)

pierre-24 committed:
Plus de 500 sur béta de profil (fix #1187)

pierre-24 committed:
Le titre des tutoriels est versionné (fix #1281)

pierre-24 committed:
Supprime des changements non-fait

pierre-24 committed:
Autorise suppression chapitre (#1189)

Ces modifications semblent permettre à nouveau la suppression de
chapitre sans problème. Cependant, les performances chutent
drastiquement, surtout sur la version online du tutoriel.

pierre-24 committed:
Amélioration des performances sur tutos

L'introduction et la conclusion ne sont plus chargés
automatiquement et doivent être chargés explicitements via
`Tutorial.load_introduction_and_conclusion()`, afin d'éviter
des lectures disques quand elles sont inutiles.

pierre-24 committed:
PEP8+documentation

pierre-24 committed:
Ajout de test unitaires

Pour :
- Assurer que la page qui liste les tutoriels en béta est bien
  accessible pour l'auteur, un utilisateur lambda et de l'extérieur
  (check #1187)
- Assurer qu'une partie et un chapitre restent visibles sur la
  version béta et en ligne malgré qu'ils aient été supprimmés
  de la version draft (check #1189)

gustavi commented on zds/tutorial/models.py:
```diff
@@ -182,34 +182,69 @@ def get_prod_path(self):
+            'source', 'sha_draft', 'sha_beta', 'sha_validation', 'sha_public']
+
+        #load functions and attributs in tree
+        for fn in fns: mandata[fn]=getattr(self,fn)()
```
Pas de for en une ligne, c'est illisible.

gustavi commented on zds/tutorial/models.py:
```diff
@@ -182,34 +182,69 @@ def get_prod_path(self):
+                args=[self.pk, mandata['slug']]
+            )
+
+        if self.in_beta() :
```
espace avant le « : » inutile.

gustavi commented on zds/tutorial/models.py:
```diff
@@ -182,34 +182,69 @@ def get_prod_path(self):
+        return mandata
+
+    def load_introduction_and_conclusion(self, mandata, sha=None, public=False) :
+        '''Explicitly load introduction and conclusion to avoid useless disk
```
C'est pas PEP257 (cf. http://legacy.python.org/dev/peps/pep-0257/#multi-line-docstrings ).

Il faut que le « ''' » soit seul sur sa dernière ligne.

artragis replied:
++ espace avant ":" inutile

gustavi commented on zds/tutorial/models.py:
```diff
@@ -182,34 +182,69 @@ def get_prod_path(self):
+        '''Explicitly load introduction and conclusion to avoid useless disk
+        access in load_dic()'''
+
+        if not public :
```
pourquoi faire un if not truc -> foo else -> bar alors que if truc -> bar else foo ?

gustavi commented on zds/tutorial/models.py:
```diff
@@ -262,16 +297,20 @@ def get_introduction(self, sha=None):
-        return intro_contenu.decode('utf-8')
+            return intro_contenu.decode('utf-8')
+
+        else :
```
le « else » ne sert à rien

gustavi commented on zds/tutorial/models.py:
```diff
@@ -290,16 +329,20 @@ def get_conclusion(self, sha=None):
-        return conclu_contenu.decode('utf-8')
+            return conclu_contenu.decode('utf-8')
+
+        else :
```
Même chose, le « else » ne sert à rien.

firm1 commented on zds/tutorial/tests.py:
```diff
@@ -1023,6 +1035,81 @@ def test_workflow_tuto(self):
+                    tuto.pk,
+                    tuto.slug,
+                    p1.pk,
+                    p1.slug]) + '?version={}'.format(sha_draft),
```
Le commentaire dit : 

> #deleted part and section HAVE TO be accessible on beta (get 200)

Et pourtant du fait un `+ '?version={}'.format(sha_draft),`

Il faudrait utiliser le sha_beta ici

pierre-24 replied:
Eeeuuuh, tu as raison et tord en même temps. En fait, c'est la variable qui est mal nommée, parce que dans le code, je me sert de ce sha pour activer la béta sur le tuto, de facto, ce sha correspond bien à sha_beta. Mais pour ne pas induire les éventuels codeurs en erreur, je vais changer la variable proprement (avec un Tutorial.objects.get())

pierre-24 replied:
... Et en fait, tu avais plus raison que tord, au vu de ce que devient `sha_draft` par la suite ^^

firm1 commented on zds/tutorial/tests.py:
```diff
@@ -1080,6 +1167,56 @@ def test_workflow_tuto(self):
+                    tuto.pk,
+                    tuto.slug,
+                    p1.pk,
+                    p1.slug]) + '?version={}'.format(sha_draft),
```
Le commentaire dit : 

> #deleted part and section still accessible on beta (get 200)

Et pourtant du fait un `+ '?version={}'.format(sha_draft),`

Il faudrait utiliser le sha_beta ici

artragis commented on zds/tutorial/views.py:
```diff
@@ -753,7 +757,8 @@ def view_tutorial_online(request, tutorial_pk, tutorial_slug):
     mandata = tutorial.load_json_for_public()
-    mandata = tutorial.load_dic(mandata)
+    mandata = tutorial.load_dic(mandata, sha=tutorial.sha_public)
+    mandata = tutorial.load_introduction_and_conclusion(mandata, public=True)
```
pourquoi doit-on assigner mandata? il n'est pas modifié à l'intérieur de la fonction?

pierre-24 committed:
Merge remote-tracking branch 'upstream/master' into fix_versionning_2

pierre-24 committed:
PEP et autres corrections mineures

- variable sha_draft→sha_beta dans
   BigTutorialTests.test_workflow_tuto
- modification de load_dic() et de load_introduction_and_conclusion()
  pour qu'elle ne renvoient plus mandata, ce qui est inutile
- retrait de plusieurs `else : return None` inutiles.

pierre-24 commented:
J'ai tenu compte des remarques effectuées sur le _review_. Comme dit Firm1, **GO QA** :)

(edit: je ne suis pas là jusque lundi prochain, donc je tiendrais compte des éventuelles remarques d'ici là. N'hésitez pas à merger le cas échéant)

artragis commented:
j'y regarderai ce week end. vue que la PR est grosse, néanmoins, on ne sera pas trop de deux à vérifier je pense.

artragis commented:
bon, je n'arrive toujours pas à faire ma QA à cause d'un problème sur mon PC.

Alex-D commented:
Je ne peux pas lancer le serveur à cause de l'erreur suivante : 

```
Unhandled exception in thread started by <function wrapper at 0x000000000347CF28>
Traceback (most recent call last):
  File "C:\django\zdsenv\lib\site-packages\django\utils\autoreload.py", line 93, in wrapper
    fn(*args, **kwargs)
  File "C:\django\zdsenv\lib\site-packages\django\core\management\commands\runserver.py", line 104, in inner_run
    now = now.decode('utf-8')
  File "C:\django\zdsenv\lib\encodings\utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeDecodeError: 'utf8' codec can't decode byte 0xfb in position 2: invalid start byte
```

Je n'avais pas cette erreur avant, donc j'imagine qu'elle est introduite par ta PR.

artragis commented:
ah, bah en fait c'était pas mon windows qui buggait.

pierre-24 commented:
Donc Alex-D et artragis, vous avez tout les deux la même erreur, que je n'ai pas (Linux?), et qui n'indique aucun fichier impliqué dans la PR, à première vue. Magnifique, ça va être drôle à débuger, ça ^^

artragis commented:
Sauf que la preprod est sur linux.
Travailles-tu avec un virtualenv @pierre-24 ?

pierre-24 commented:
Non. Mais dans tout les cas, on peut pas laisser passer un machin qui fait planter les Windows des dev' sous prétexte que la prod' est sous Linux ;)

(EDIT: j'imagine que ça le fait qu'avec ma PR ?)

artragis commented:
Ce que je veux dire c'est que ça plante **aussi** sur la preprod qui est
pourtant sur Linux.

Le 7 août 2014 09:55, Pierre Beaujean notifications@github.com a écrit :

> Non. Mais dans tout les cas, on peut pas laisser passer un machin qui fait
> planter es Windows des dev' sous prétexte que la prod' est sous Linux ;)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1322#issuecomment-51441712
> .

artragis commented:
> Non.

Je pense que c'est là qu'il y a un "hic", ton env de dev doit être plus
complet que le virtualenv qu'on nous demande d'installer. Du coup tu as
peut être moins de soucis d'utf8

Le 7 août 2014 10:04, francois dambrine artragis@gmail.com a écrit :

> Ce que je veux dire c'est que ça plante **aussi** sur la preprod qui est
> pourtant sur Linux.
> 
> Le 7 août 2014 09:55, Pierre Beaujean notifications@github.com a écrit :
> 
> Non. Mais dans tout les cas, on peut pas laisser passer un machin qui fait
> 
> > planter es Windows des dev' sous prétexte que la prod' est sous Linux ;)
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/zestedesavoir/zds-site/pull/1322#issuecomment-51441712
> > .

pierre-24 commented:
@artragis : l'erreur qu'a donné Alex-D est bien une erreur Windows aussi. Quand à lapréprod', elle se porte bien, dieu merci. 

> Je pense que c'est là qu'il y a un "hic", ton env de dev doit être plus
> complet que le virtualenv qu'on nous demande d'installer. Du coup tu as
> peut être moins de soucis d'utf8

Possible. Sous Linux, on est moins affecté par les erreurs d'encodage ... Ce qui ne m'aide pas dans ce cas-ci :(

pierre-24 committed:
Merge branch 'master' of https://github.com/zestedesavoir/zds-site into fix_versionning_2

pierre-24 commented:
Bon, ça m'ennuie parce que je n'arrive pas à trouver la cause de cette erreur sur internet, et je n'arrive pas à la régler non plus, puisque je n'arrive pas à la reproduire. Seule piste qui ressemble un peu à quelque chose que j'ai : quel est la commande (exacte) que vous employez pour lancer django ? Pouvez vous réessayer avec un _host_ explicite (`python manage.py runserver 0.0.0.0:8000`) et voir si l'erreur ce produit toujours ?

C'est ça ou ... Ou je vois pas, en fait --" (je vais tenter d'aller taper des `u` un peu partout, pour voir)

pierre-24 committed:
Tentative correction unicode

?!?

pierre-24 commented:
et de manière générale : https://docs.djangoproject.com/en/dev/ref/unicode/

(mais ça n'as jamais posé problème avant :o )

artragis commented:
La commande que je tappe est : `C:\Python27\python.exe .\manage.py
runserver`

2014-08-07 10:53 GMT+02:00 Coveralls notifications@github.com:

> [image: Coverage Status] https://coveralls.io/builds/1054240
> 
> Coverage increased (+0.22%) when pulling _5a19e34
> https://github.com/zestedesavoir/zds-site/commit/5a19e3429c8b71da3799350b6cf91b8b39eec38c
> on pierre-24:fix_versionning_2_ into _aaa5aea
> https://github.com/zestedesavoir/zds-site/commit/aaa5aea1aea6f62574f9323eafdc7d88d44e69bd
> on zestedesavoir:master_.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1322#issuecomment-51446650
> .

pierre-24 commented:
@artragis : 

quand tu saura, teste alors C:\Python27\python.exe .\manage.py runserver 0.0.0.0:8000`. Je suis pas convaincu que ça changera quelque chose, mais je l'ai vu à plusieurs reprise, donc autant essayer. Avant de faire ça, teste comme tu fais d'habitude, j'ai mis un commit qui règle pour moi la (les?) seule(s) erreur(s) UTF-8 possible dans ma PR :s

gustavi commented on zds/tutorial/models.py:
```diff
@@ -182,34 +182,69 @@ def get_prod_path(self):
+
+        #load functions and attributs in tree
+        for fn in fns: 
+            mandata[fn]=getattr(self,fn)()
```
À quoi servent les parenthèses à la fin ?

pierre-24 replied:
A rien, j'imagine ^^ (j'ai des relents de C quand je code, donc quand j'ai vu ça chez Firm1, j'ai pas tilté)

gustavi replied:
Je vais enculer les mouches mais ton code n'a pas toujours la même syntaxe dans 
: 

```
    fns = [u'is_big', u'is_mini', u'have_markdown',u'have_html',
        u'have_pdf', u'have_epub', u'get_path']
    attrs = [u'pk',u'authors',u'subcategory',u'image',u'pubdate', u'update',
        u'source', u'sha_draft', u'sha_beta', u'sha_validation',
        u'sha_public']
```

Des fois il y a une espace avant la virgule et des fois non (une espace étant ce qu'il faut), ce serait bien d'homogénéisé tout ça ;)

gustavi replied:
Pareil pour 

```
mandata[fn]=getattr(self,fn)
```

qu'il faudrait remplacer par 

```
mandata[fn] = getattr(self, fn)
```

firm1 replied:
> A rien, j'imagine ^^

Mauvaise réponse ^^

pierre-24 replied:
Ben oui, mais quand je les enlève, ça change rien ;) ... Donc ?

pierre-24 committed:
Retrait de syntaxe inutile

(suite à la remarque de gustavi)

cgabard commented:
vu que le model est modifié, il ne faut pas faire un migrate ou un truc du genre ? 

En attendant je vais tenter de refaire un env propre pour voir si je le reproduit en local (mais je suis sous linux)

cgabard commented:
Environnement neuf sous linux, impossible de reproduire le bug

artragis commented:
> vu que le model est modifié, il ne faut pas faire un migrate ou un truc du
> genre ?

A noter : il me semble avoir fait un migrate justement. Je regarde ça se
soir, mais AMHA, c'est un problème purement windows.

2014-08-07 15:45 GMT+02:00 Christophe Gabard notifications@github.com:

> Environnement neuf sous linux, impossible de reproduire le bug
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1322#issuecomment-51473074
> .

gustavi commented:
Je crois que @cgabard a raison, il manque les fichiers de migration !

pierre-24 commented:
> Je crois que @cgabard a raison, il manque les fichiers de migration !

... Non ? Je veux dire, le modèle est modifié, mais uniquement son code, j'ai pas ajouté ou retiré de colonne dans la base ou autre, j'ai juste repensé la manière de traiter les machins. Je doit quand même créer un fichier de migration ?

> A noter : il me semble avoir fait un migrate justement. Je regarde ça se
> soir, mais AMHA, c'est un problème purement windows.

C'est un problème purement Windows, j'en suis sur et certain, et je m'en excuse. Même question : moi, j'ai jamais créé de fichier de migration, je devrais ?

Alex-D commented:
Nop, si t'as pas touché au schéma du model, pas besoin de faire de fichier de migration.

Pour la commande, je lance toujours comme ça : 

`python manage.py runserver 0.0.0.0:3000` et ça a toujours fonctionné.

SpaceFox commented:
Je confirme le problème, sur un environnement Windows tout neuf...

SpaceFox commented:
Le problème **ne vient pas** de la PR, j'ai le même sur `master` !

Ou alors il porte sur quelque chose qui n'est pas réinitialisé quand on change de branche et qu'on supprime tous les *.pyc... je me demande si le problème ne vient pas en réalité d'une dépendance qu'on aurait mise à jour récemment.

SpaceFox commented:
Comme le bug d'encodage Windows est présent sur la version en prod (tag v1.0-RC7), on peut QAtiser cette branche, parce qu'elle corrige quand même un problème très très chiant en prod...

SpaceFox commented:
Il y a des combines pour tester ça sous Windows, cf le ticket #1353 :)

Alex-D commented:
A la limite, faut mettre cette branche sur la preprod pour pouvoir tester à plusieurs en même temps en conditions réelles et voir si rien ne plante comme sur la prod actuelle.

pierre-24 commented:
Ça me rassure de savoir que le problème ne vient pas de moi :) (et le rapport de Spacefox est excellentissime, j'adore).

Bon, qui sous Linux pour faire une bonne QA ? Parce que la PR périme :p

SpaceFox commented:
La préprod est [HS pour le moment](http://zestedesavoir.com/forums/sujet/980/serveur-de-preprod-vas-surement-etre-down/?page=1#p15550)

SpaceFox commented:
Je vais QA cette PR.

SpaceFox commented:
La QA passe pour les big-tutos.

Elle ne passe pas pour les mini-tutos : je peux créer un mini-tuto, mais pas ajouter d'extrait.

Si je clique sur "ajouter un extrait", je me prends cette erreur : 

```
Environment:


Request Method: GET
Request URL: http://localhost:8000/tutoriels/nouveau/extrait/?chapitre=

Django Version: 1.6.5
Python Version: 2.7.5
Installed Applications:
('django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.sites',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'django.contrib.sitemaps',
 'django.contrib.humanize',
 'easy_thumbnails',
 'easy_thumbnails.optimize',
 'south',
 'crispy_forms',
 'email_obfuscator',
 'haystack',
 'munin',
 'zds.utils',
 'zds.pages',
 'zds.gallery',
 'zds.mp',
 'zds.article',
 'zds.forum',
 'zds.tutorial',
 'zds.member',
 'django.contrib.admin',
 'debug_toolbar')
Installed Middleware:
(u'debug_toolbar.middleware.DebugToolbarMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'zds.utils.ThreadLocals',
 'zds.middlewares.SetLastVisitMiddleware.SetLastVisitMiddleware')


Traceback:
File "/home/spacefox/virtualenvs/zdsenv/local/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  112.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/home/spacefox/virtualenvs/zdsenv/local/lib/python2.7/site-packages/django/contrib/auth/decorators.py" in _wrapped_view
  22.                 return view_func(request, *args, **kwargs)
File "/home/spacefox/git/zds-site/zds/tutorial/views.py" in add_extract
  1858.         chapter_pk = int(request.GET["chapitre"])

Exception Type: ValueError at /tutoriels/nouveau/extrait/
Exception Value: invalid literal for int() with base 10: ''
```

Sur ta branche, base neuve, Linux.

pierre-24 commented:
Merci pour cette QA, je vais faire un correctif :) (m'as tout l'air d'être un id qui manque dans l'url)

pierre-24 committed:
Merge branch 'master' of https://github.com/zestedesavoir/zds-site into fix_versionning_2

pierre-24 committed:
Fix édition mini-tuto (+typo)

pierre-24 commented:
Voilà, c'est corrigé (et j'en ai profité pour tenir compte des remarques de gustavi)

SpaceFox commented:
Je QA ça.

SpaceFox commented:
QA OK. Bon boulot, merci !

Je MEP ce soir, si tu as le temps de dégager le commentaire inutile c'est parfait ; sinon tant pis :(

pierre-24 commented:
Je te dégage ça de suite :)

pierre-24 committed:
Retrait d'un commentaire inutile
